### ISSUE TITLE
Low precision and mAP on custom data

### ISSUE THREAD

**Issue by FaizahMushtaq** [ultralytics/yolov5#3617]
## ❔Question

Hello @glenn-jocher, I'm training yolov5 for my custom dataset consist of two classes Normal and depressed, the images are basically spectrograms, when I trained it for 4k images, 2k belonging to each class I achieved mAP_0.5  0.996 with 300 epochs from scratch. Now training same data with 7k images 3500 Normal and 3900 depressed the mAP is 0.52 and precision remains 0.5 even after 600 epochs and 1000. It increases till 0.54 but at end result is 0.52 and precision remains constant . Can  you please help me to find out what's going wrong?
Here are training images
![de](https://user-images.githubusercontent.com/62736878/121999081-4a60ab00-cdc6-11eb-93e5-c46ad667d9e6.png)
and  training
![e](https://user-images.githubusercontent.com/62736878/121999163-6401f280-cdc6-11eb-9311-fb4333e34fec.png)

**Comment by glenn-jocher** [comment]
@FaizahMushtaq 👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by FaizahMushtaq** [comment]
I have followed all the steps, but still there is no improvement ,tried  with 6k dataset the mAP and precision still 0.52.

![7kyolov5](https://user-images.githubusercontent.com/62736878/122172357-1ce53180-ce9a-11eb-94d9-d272e812a48e.png)

![Screenshot 2021-06-16 120121](https://user-images.githubusercontent.com/62736878/122172986-ba406580-ce9a-11eb-85dc-c400fac94179.png)

**Comment by GrimReaperSam** [comment]
Are you solving a classification problem with a detection model?

**Comment by FaizahMushtaq** [comment]
@GrimReaperSam Yes

**Comment by GrimReaperSam** [comment]
You shouldn't! You should just use a classification model to do that.
A lot of your training effort is wasted on identifying objectness and bounds when they are not relevant to you.

**Comment by FaizahMushtaq** [comment]
But it was giving mAP 0.996 now after 1 epoch the recall becomes 1 and precision, mAP remain constant what is reason behind it?

**Comment by GrimReaperSam** [comment]
You give too little information to figure this out..

**Comment by FaizahMushtaq** [comment]
@GrimReaperSam I'm using yolov5s with all default parameters batch size 16, img size 416 and for 300 epochs, the data set consist of 7000 images, 4900 for training 2450 belong to each normal and depressed class. 1400 for validation .
![image](https://user-images.githubusercontent.com/62736878/122541370-ff070080-d042-11eb-9201-55de830bacc1.png)

Then after 1 epochs, the recall reaches 1 and mAP and precsion 0.5, this behaviour remain constant till all epochs completed even I run it for 1000 epochs.

![image](https://user-images.githubusercontent.com/62736878/122541680-5ad18980-d043-11eb-9189-12b2ad84f5f0.png)
I tried for 4000 images but same results recall is 1 even at start

![image](https://user-images.githubusercontent.com/62736878/122560412-038ae380-d05a-11eb-82d6-8a19fcbe8043.png)

**Comment by ChinaRush** [comment]
your train_loss is shaking,maybe you can adjust your learning rate or batch-size? or there may be something wrong of your data
